### ISSUE TITLE
Windows installer fails if MDSplus dll's are in use (i.e. mdsip server is enabled)

### ISSUE THREAD

**Issue by tfredian** [MDSplus/mdsplus#1704]
I may be wrong but I think this used to work. When MDSplus is updated and the system had a running mdsip service the MDSplus dll's being used by that service are not updated. Since the MDSplusVersion mechanism changed mdsdcl cannot find the data structure used in the mdsshr.dll so it fails to start. Similarly the python module cannot find the required structure and complains when importing MDSplus. If the mdsip service is stopped and you deinstall MDSplus and then reinstall it, it seems to fix the problem.

The original scheme which I thought worked was to rename the dll's that were in use since they cannot be removed and then install the new dll's. I did find some renamed dll's in system32 so perhaps that happened as expected but maybe Windows is now using the renamed dll's until all MDSplus processes exit. If this is the case then simply restarting the mdsip service would fix the problem. I did not try this. I don't know what the standard way of dealing with this in windows installers is. If the rename no longer works properly then perhaps the installer needs to detect dll's actively in use and stop the processes using those dll's during the deinstall of MDSplus.

**Comment by zack-vii** [comment]
Well the designed workaround this problem is to setup a postinstall and make the user restart the system. Stopping the mdsip server before installing should also do the trick. What I found was that the installer just fails to move the dlls from the install folder to system32. In that case you need to manually stop the services or all running mdsplus binaries in general and move the remaining files.
If renaming fixes it, it is still a very dirty solution. as those files will likely not be removed by the uninstaller.
we should implement the following
https://blogs.msdn.microsoft.com/windows_installer_team/2005/09/18/insights-into-the-underlying-windows-installer-reboot-behavior/

**Comment by zack-vii** [comment]
Since we alway run uninstall before installing a new version we should force the uninstaller to remove&rename files thoroughly, as intended by microsoft.

**Comment by zack-vii** [comment]
looks like we could make use of the [LockedList](https://nsis.sourceforge.io/LockedList_plug-in) plugin for nsis. It would pop a dialog if files cannot be deleted.

**Comment by tfredian** [comment]
The LockedList plugin looks promising. We could also probably stop/start the mdsip services when updating:

```
sc stop "MDSplus 8000 >nul"
sc stop "MDSplus 8100 >nul"
```
invoke LockedList plugin (perhaps in a loop until nothing is locked or the uninstall is canceled?)
uninstall/remove files
install mdsplus
```
sc start "MDSplus 8000 >nul"
sc start "MDSplus 8100 >nul"
```

**Comment by zack-vii** [comment]
I was playing with that plug in just now and it is very clumsy and slow. i will try the windows method of renaming the files and marking them for removal. if /reboot  is activated nsis will put files on a list and schedule them for removal. however if we wont rename them before a reboot may break a following installation. We would not want to force the user to reboot on each update. or we pop a dialog whenever a file cannot be deleted and ask the user to close any open process that may access it. the user could then decide to retry or to ignore. ignoring would trigger the attempt to rename. the process could be an mdsip but may also be a python session. so focusing on the default mdsip services is only half the way.

**Comment by tfredian** [comment]
I just suggested turning off the services in case we used LockList since LockList would probably not provide different info regarding the need to stop a service vs stopping various running applications. Perhaps a better solution is to go back to the rename approach with the addition of a delete of renamed files after a reboot. Running applications would probably continue to work ok and newly started applications would use the new version. Possibly renaming locked files to a mdsplus-locked subdirectory and adding a deletion of the subdirectories after a reboot would be simpler. We probably wouldn't even need any notification popup unless we wanted to notify the user they would need to restart the application to use the upgraded version.

**Comment by zack-vii** [comment]
the uninstaller attempts to do something with the  services
nsExec::ExecToLog /OEM /timeout=5000 '"$SYSDIR\mdsip_service.exe" "-r -p 8100"'

**Comment by tfredian** [comment]
I suspect that only attempts to remove the service definition which probably doesn't succeed if the service is running at the time. I'm not sure if we can detect if this uninstall is being done because MDSplus is being updated or just deinstalled. If we can detect probably we should just stop the services if and update is in progress and stop and remove the services if just doing a deinstall.

**Comment by zack-vii** [comment]
should be resolved as part of #1705
